nixos/modules/services/torrent/transmission.nix

```diff
@@ -100,56 +115,180 @@ in
+
+    assertions = [
+      { assertion = builtins.match "^/.*" cfg.home != null;
+        message = "`services.transmission.home' must be an absolute path.";
+      }
+      { assertion = types.port.check cfg.settings.rpc-port;
+        message = "${toString cfg.settings.rpc-port} is not a valid port number for `services.transmission.settings.rpc-port`.";
+      }
+      # In case both port and settings.rpc-port are explicitely defined: they must be the same.
+      { assertion = !options.services.transmission.port.isDefined || cfg.port == cfg.settings.rpc-port;
+        message = "`services.transmission.port' is not equal to `services.transmission.settings.rpc-port'";
+      }
     ];
 
+    services.transmission.settings =
+      optionalAttrs options.services.transmission.port.isDefined { rpc-port = cfg.port; };
+
     systemd.services.transmission = {
       description = "Transmission BitTorrent Service";
       after = [ "network.target" ] ++ optional apparmor "apparmor.service";
-      requires = mkIf apparmor [ "apparmor.service" ];
+      requires = optional apparmor "apparmor.service";
       wantedBy = [ "multi-user.target" ];
+      environment.CURL_CA_BUNDLE = etc."ssl/certs/ca-certificates.crt".source;
+      preStart = ''
+        set -eux
+        ${pkgs.jq}/bin/jq --slurp add ${settingsFile} '${cfg.credentialsFile}' >'${stateDir}/${settingsDir}/settings.json'
+      '';
 
-      # 1) Only the "transmission" user and group have access to torrents.
-      # 2) Optionally update/force specific fields into the configuration file.
-      serviceConfig.ExecStartPre = preStart;
-      serviceConfig.ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f --port ${toString config.services.transmission.port} --config-dir ${settingsDir}";
-      serviceConfig.ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
-      serviceConfig.User = cfg.user;
-      serviceConfig.Group = cfg.group;
-      # NOTE: transmission has an internal umask that also must be set (in settings.json)
-      serviceConfig.UMask = "0002";
+      serviceConfig = {
+        WorkingDirectory = stateDir;
+        ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        User = cfg.user;
+        Group = cfg.group;
+        StateDirectory = removePrefix "/var/lib/" stateDir + "/" + settingsDir;
+        StateDirectoryMode = "0700";
+        BindPaths =
+          optional (cfg.home != stateDir) "${cfg.home}/${settingsDir}:${stateDir}/${settingsDir}"
+          ++ [ "${cfg.settings.download-dir}:${stateDir}/Downloads" ]
+          ++ optional cfg.settings.incomplete-dir-enabled "${cfg.settings.incomplete-dir}:${stateDir}/.incomplete";
+        # The following options give:
+        # systemd-analyze security transmission
+        # → Overall exposure level for transmission.service: 1.5 OK
+        AmbientCapabilities = "";
+        CapabilityBoundingSet = "";
```

Comment:
Tiny nitpick, but I **believe** `[]` is preferred here... (someone correct me if I'm wrong).

Reply:
Using `[]` would not work here because it would not put the entry at all, but an empty entry is needed to disable the capabilities. If extensibility is a concern (but AFAIU it's not a concern in this case), `[""]` would do, since for instance `["", "CAP_NET_ADMIN"]` would disable all capabilities but `CAP_NET_ADMIN`.

Reply:
@ju1m thanks for mentioning, oops!